blocks/search.py

```diff
@@ -0,0 +1,266 @@
+                 attended_mask, inputs_dict):
+        self.beam_size = beam_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.attended_mask = attended_mask
+        self.inputs_dict = inputs_dict
+        self.generate_names = sequence_generator.generate.outputs
+        self.init_computer = None
+        self.next_computer = None
+        self.attended_computer = None
+        self.initial_state_computer = None
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+        self.compiled = False
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beam search."""
+        generator = self.sequence_generator
+        attended = self.attended
+        attended_mask = self.attended_mask
+
+        self.attended_computer = function(self.inputs_dict.values(),
+                                          [attended, attended_mask],
+                                          on_unused_input='ignore')
+
+        initial_states = OrderedDict()
+        for name in self.state_names:
+            initial_states[name] = generator.initial_state(
+                name,
+                self.beam_size,
+                attended=attended)
+
+        self.initial_state_computer = function([attended, attended_mask],
+                                               initial_states.values(),
+                                               on_unused_input='ignore')
+
+        # Define inputs for next values computer
+        cur_variables = OrderedDict()
+        for name, value in initial_states.iteritems():
+            cur_value = tensor.zeros_like(value)
+            cur_value.name = name
+            cur_variables[name] = cur_value
+
+        next_generated = generator.generate(attended=attended,
+                                            attended_mask=attended_mask,
+                                            iterate=False,
+                                            n_steps=1,
+                                            batch_size=self.beam_size,
+                                            return_dict=True,
+                                            **cur_variables)
+        cg_step = ComputationGraph(next_generated.values())
+        next_probs = VariableFilter(
+            application=generator.readout.emitter.probs,
+            name='output')(cg_step.variables)[-1]
+        # Create theano function for next values
+        self.next_computer = function([attended, attended_mask,
+                                       cur_variables['states']],
+                                      next_generated.values() + [next_probs])
+        self.compiled = True
+
+    @construct_dict
+    def compute_contexts(self, inputs_dict):
+        contexts = self.attended_computer(*inputs_dict.values())
+        return ["attended", "attended_mask"], contexts
```

Comment:
I think that `OrderedDict(zip(*))` is simple enough and does not justify using a decorator. It is a very well-known snippet as well.